Complex schedules UI

keithjgrant opened the pull request:
##### SUMMARY
Adds ability to set more complex patterns in schedule - multiple recurring patterns can be built together (e.g. an hourly AND a daily recurrence)

- ability to add exceptions is forthcoming (e.g. run daily EXCEPT on certain dates) (#12542)

##### ISSUE TYPE
 - New or Enhanced Feature

##### COMPONENT NAME
 - UI

keithjgrant committed:
refactor ScheduleFormFields into own file

keithjgrant committed:
refactor ScheduleForm

keithjgrant committed:
wip complex schedules form

keithjgrant committed:
build rruleset from inputs

keithjgrant committed:
update schedule form validation for multiple repeat frequencies

keithjgrant committed:
add basic rrule set parsing when opening schedule form

keithjgrant committed:
complex schedule bugfixes, handle edge cases, etc

keithjgrant committed:
fix schedule saving/parsing for single-occurrence schedules

keithjgrant committed:
working with timezone issues

keithjgrant committed:
fix rrule until times to be in UTC

keithjgrant committed:
update tests for new schedule form format

keithjgrant committed:
update ouiaIds

keithjgrant committed:
tweak schedules spacing

keithjgrant committed:
update ScheduleForm tests

keithjgrant commented on awx/ui/src/components/Schedule/shared/ScheduleForm.test.js:
```diff
@@ -1096,40 +1233,64 @@ describe('<ScheduleForm />', () => {
       expect(
         wrapper.find('TimePicker[aria-label="End time"]').prop('value')
-      ).toBe('1:00 AM');
+      ).toBe('12:00 AM');
```
I think this was a timezone/DST bug in the previous version...? America/New_York should be UTC-5:00 in January

keithjgrant commented:
~~This has some minor cleanup remaining, but~~ I'd like to get some eyes on it now. The logic to add "Exceptions" fields is still TODO in a follow-up PR (and a little bit of it is here now, just commented out)

keithjgrant committed:
show message for unsupported schedule types

keithjgrant committed:
default schedules to browser timezone

AlexSCorey commented:
@keithjgrant I've started to look at this.  First, this work is amazing and is a great leap forward in helping customers automate more.  

I did want to note 1 thing:

- [ ] If you create and turn on a survey and create and turn on a schedule, then return to the survey to turn it off, and then return to edit the schedule you cannot edit the schedule without clicking on prompt.  This is a problem, because the survey is off and should not impede the ability to save/edit a schedule
-[ ] The repeat frequency field keeps track of how many items the user has clicked on, but now which items those are.  Shouldn't we show chips there? 
![Screen Shot 2022-07-20 at 2 13 05 PM](https://user-images.githubusercontent.com/39280967/180053264-93eecf3d-a15f-4bc4-b6d5-cd6b7925c71d.png)
- [x] Also looks like Repeat Frequency label is not marked for translated
![Screen Shot 2022-07-20 at 2 22 29 PM](https://user-images.githubusercontent.com/39280967/180054890-46bc55c6-9bb6-4c2e-b11e-0241f7f97ec1.png)
- [ ] I'd be cautious about using abbreviations for the days of the week because they may not translate properly
- [x] This frequency schedule (weekly and monthly, with repeats on both) results in a `Repeat frequency` of `yearly` on the details view, with the first and next run fields containing the wrong value.
![Screen Shot 2022-07-20 at 2 24 39 PM](https://user-images.githubusercontent.com/39280967/180056021-e8c94366-02ff-46f2-bb0f-7b192a48c0a9.png)
![Screen Shot 2022-07-20 at 2 22 29 PM](https://user-images.githubusercontent.com/39280967/180056024-b09187bd-6fdd-450b-9ceb-e9bb80979b7d.png)

akus062381 commented:
feedback after manual testing:

- [ ] some time zones are showing the following text upon being selected on the form: `Warning: {selectedValue} is a link to {0} and will be saved as that.`
- [x] some time zones show up one way when selected in the form but change after the user saves the form. Example: America/Toronto saves and shows up as Canada/Eastern
- [x] consider providing more explicit information on the schedules detail page about what the user stipulated on the schedule form
- [x] in the form field on the following snippet, I feel like a better user experience would be for actual names of intervals to show up (minute, hour, etc)  instead of numbers 
![image](https://user-images.githubusercontent.com/63685497/180075561-6e9c775f-4c78-4ad7-aa48-0bf0ec6b1197.png)

kialam commented:
I'm seeing some errors in the console that we'll probably want to resolve. When trying to edit a management job (5, Cleanup Expired OAuth 2 Tokens https://localhost:3001/#/management_jobs/5/schedules/5)

1. When you select the repeat frequency dropdown I get this error:
<img width="1027" alt="Screen Shot 2022-07-20 at 2 26 25 PM" src="https://user-images.githubusercontent.com/2293210/180085319-120376a6-e1fe-4996-927e-da06d25f8be6.png">

2. I cannot select a day when trying to create a weekly frequency and get this error:
<img width="1013" alt="Screen Shot 2022-07-20 at 2 24 21 PM" src="https://user-images.githubusercontent.com/2293210/180085367-fbb2f046-4564-4a21-9765-2a10cb07722f.png">

kialam commented on awx/ui/src/components/Schedule/shared/ScheduleFormFields.js:
```diff
@@ -0,0 +1,194 @@
+        }
+        label={t`Repeat frequency`}
+      >
+        <FrequencySelect
```
When selecting multiple frequencies the `Repeat Frequency` detail in the Schedule Detail page may be a bit misleading:
<img width="573" alt="Screen Shot 2022-07-20 at 3 18 19 PM" src="https://user-images.githubusercontent.com/2293210/180092584-6cdc4c2f-d506-4096-9ca8-28f09d40b4de.png">

I have a schedule that repeats every min and hour, however the `Repeat Frequency` detail reads "every year."

keithjgrant committed:
show error type/message in ErrorDetail

keithjgrant committed:
shows frequencies on ScheduleDetails view

keithjgrant committed:
handles nullish values

keithjgrant commented:
Changes up and ready for review. I've updated the Details page with a full schedule breakdown for run frequencies. Some other thoughts/replies to specific comments below:

> * If you create and turn on a survey and create and turn on a schedule, then return to the survey to turn it off, and then return to edit the schedule you cannot edit the schedule without clicking on prompt.  This is a problem, because the survey is off and should not impede the ability to save/edit a schedule

@AlexSCorey This appears to be an existing bug in devel. I filed #12585 to track it

> * Also looks like Repeat Frequency label is not marked for translated

@AlexSCorey it is marked, but I haven't run extract-strings and compile-strings, so this should get picked up when we run them for the next release

> * The repeat frequency field keeps track of how many items the user has clicked on, but now which items those are.  Shouldn't we show chips there?

@AlexSCorey I tend to agree, though the current approach matches mockups & patternfly component for multiselect. We can definitely discuss/revisit later


> * some time zones are showing the following text upon being selected on the form: `Warning: {selectedValue} is a link to {0} and will be saved as that.`

@akus062381 This is a translation/string replacement issue. These are replacing normally in my local environment, though. Any idea why they would be different in your instance?


> I'm seeing some errors in the console that we'll probably want to resolve. When trying to edit a management job (5, Cleanup Expired OAuth 2 Tokens https://localhost:3001/#/management_jobs/5/schedules/5)

@kialam I couldn't reproduce these, though I suspect they are resulting from a particular format of a schedule created in previous version. I updated those two lines to handle undefined/null values.

mabashian reviewed (approved):
I'm going to approve this based on my manual testing.  I think the UX is definitely a vast improvement and I'm excited to be expanding the scheduler functionality for the community to cover more cases.  I'll defer to @AlexSCorey and @kialam on the actual code changes.

InnocentK reviewed (approved):
Great feature Keith! This looks much better now with the fixes you pushed. From QE side I think this is good to go now. Ignore those failing tests--a lot of those are just updates QE needs to make to our E2E tests to accommodate the changes from this PR. We should be pushing fixes to those failures in the near future

kialam reviewed (approved):
lgtm
